modnet/sklearn.py

```diff
@@ -0,0 +1,134 @@
+
+    def fit(self, X, y=None):
+        """Probably not needed except for some matminer featurizers (e.g. I think SOAP needs to be fitted before)."""
+        return self
+
+    def transform(self, X, y=None):
+        """Transform the input data (i.e. containing the composition and/or structure to the features"""
+
+    @classmethod
+    def from_preset(cls, preset):
+        """Initializes a MODNetFeaturizer class based on some preset.
+
+        Args:
+            preset: Name of the preset (e.g. "DeBreuck2020" could trigger the Structure+Composition featurizers)
+
+        Notes:
+            See in matminer how it is done, e.g. SiteStatsFingerprint.from_preset method.
+        """
+
+
+class RR(TransformerMixin, BaseEstimator):
+    """Relevance-Redundancy (RR) analysis.
+
+    Feature selection and ordering procedure based on the relevance-redundancy algorithm developed in ARTICLE.
+
+    Blabla
+
+    Notes:
+        1. Uses preprocessing.get_features_relevance_redundancy under the hood.
+        2. To check if it is ok to have y that is *NOT* ignored (i.e. not y=None) in the fit method. Usually a
+           Transformer's fit method signature is fit(self, X, y=None, ...).
+    """
+
+    def __init__(self, n_feat: Union[None, int]=None, rr_parameters: Union[None, Dict]=None):
+        """Constructor for RR transformer.
+
+        Args:
+            n_feat: Number of features to keep and reorder using the RR procedure (default: None, i.e. all features).
+            rr_parameters: Allow to tune p and c parameters. (default: None, i.e. use the dynamical setting in ARTICLE).
+
+        Notes:
+            This method should not be changed
+        """
+        self.n_feat = n_feat
+        self.rr_parameters = rr_parameters
+
+    def fit(self, X, y, nmi_feats_target=None, cross_nmi_feats=None):
+        """Fit the model with X and y.
+
+        Allow to accept numpy arrays for X and y or pandas Dataframes ?
+
+        Args:
+            X: Training input data
+            y: Training output data
+            nmi_feats_target: If a precomputed nmi features=>target is available.
+            cross_nmi_feats: If a precomputed cross-nmi features=>features is available.
+        """
+
+    def transform(self, X, y=None):
+        """Transform the inputs X based on a fitted RR analysis.
+
+        Notes:
+            I don't remember if the signature should be transform(self, X, y=None) or transform(self, X).
+
+        Args:
+            X: Input data
+            y: Output data (should be ignored)
+
+        Returns:
+            Filtered and ordered data according to the fitted RR analysis.
+        """
+
+
+class MODNet(RegressorMixin, BaseEstimator):
+    """MODNet model.
+
+    Blabla.
+
+    Notes:
+        No assumption on the features here, just a list of numbers.
+        What makes a MODNet model special with respect to using directly keras ? I would say that it is always doing
+        some joint learning, maybe something else ?
```

Comment:
Indeed, MODNet class is a wrapper around Keras. The added value is that a user can create a neural network in a single line, without dealing with the Keras API. This is somewhat convenient for single target learning, and very needed for multi-target learning.

Reply:
That is what I thought. I think it makes sense. I would try to make it slightly more flexible, i.e. allowing to have between 2 and N blocks, instead of being fixed to 4 blocks. Maybe also add options to regularize ?

Reply:
I think the hierarchical multi-target architecture in the final layers is quite unique to MODNet, too. Do you think calculating cross NMI between targets could provide a way to automatically generate the optimal architecture? I wanted to have a play with this but haven't had time yet

Reply:
This is a very interesting idea, depending on the similarity, more or less layers/neurons should be shared...